lua/nvim-treesitter/install.lua

```diff
@@ -14,44 +14,80 @@ local repositories = {
   }
 }
 
--- TODO(kyazdani): there might be a better way to get the plugin path
-for _, path in pairs(api.nvim_list_runtime_paths()) do
-  if string.match(path, '.*/nvim%-treesitter') then
-    package_path = path
-    break
+local function get_package_path()
+  for _, path in pairs(api.nvim_list_runtime_paths()) do
+    if string.match(path, '.*/nvim%-treesitter') then
+      return path
+    end
   end
+
+  return nil
+end
+
+local function get_cache_dir()
+  local home = fn.get(fn.environ(), 'HOME')
+  local xdg_cache = fn.get(fn.environ(), 'XDG_CACHE_HOME')
+
+  if xdg_cache == 0 then
+    xdg_cache = home .. '/.cache'
+  end
+
+  if luv.fs_access(xdg_cache, 'RW') then
+    return xdg_cache
+  elseif luv.fs_access('/tmp', 'RW') then
+    return '/tmp'
+  end
+
+  return nil
 end
 
 local function create_compile_command(ft, path, files)
-  -- TODO(kyazdani): should download the parser in $XDG_CACHE_HOME instead of /tmp
-  return "cd /tmp/tree-sitter-"..ft..[[ && \
-    gcc -o parser.so -shared ]]..files..[[ -Os -I./src && \
-    mv parser.so ]]..path.."/parsers/"..ft..[[.so && \
-    rm -rf /tmp/tree-sitter-]]..ft
+  local cache_folder = get_cache_dir()
+  if not cache_folder then
+    return 'Invalid cache rights, $XDG_CACHE_HOME or /tmp should be read/write', nil
+  end
+
+  local project_repo = cache_folder..'/tree-sitter-'..ft
+  local compile_command = "gcc -o parser.so -shared "..files.." -Os -I./src"
+  local parser_lib = path.."/parser/"..ft..".so"
+
+  return nil, string.format("cd %s && %s && mv parser.so %s && rm -rf %s", project_repo, compile_command, parser_lib, project_repo)
 end
 
 local function create_download_command(url)
-  return "cd /tmp && git clone "..url
+  local cache_folder = get_cache_dir()
+  if not cache_folder then
+    return 'Invalid cache rights, $XDG_CACHE_HOME or /tmp should be read/write', nil
+  end
+
+  return nil, string.format("cd %s && git clone %s", cache_folder, url)
```

Comment:
We should check if git is available here, as this may cause errors afterwards

Reply:
Ok so, i started rewriting some code to use jobs instead of system so the command becomes non-blocking, then i'll be able to detect job errors.

Reply:
also i checked if git is an executable before, but i'll move it inside this function to make it clearer. Also i'm refactoring theses parts of the code to make it easier to understand.